jigglypuffRL/common/trainer.py

```diff
@@ -0,0 +1,203 @@
+                 render=False, max_ep_len=1000, distributed=False, ckpt_log_name='experiment',
+                 steps_per_epoch=4000, epochs=10, device='cpu', log_interval=10, batch_size=50,
+                 seed=None):
+        self.agent = agent
+        self.env = env
+        self.logger = logger
+        self.off_policy = off_policy
+        if self.off_policy and buffer==None:
+            if self.agent.replay_buffer is None:
+                raise Exception("Off Policy Training requires a Replay Buffer")
+            else:
+                self.buffer = self.agent.replay_buffer
+        self.save_interval = save_interval
+        self.render = render
+        self.max_ep_len = max_ep_len
+        self.ckpt_log_name = ckpt_log_name
+        self.steps_per_epoch = steps_per_epoch
+        self.epochs = epochs
+        self.device = device
+        self.log_interval = log_interval
+        self.batch_size = batch_size
+        if seed is not None:
+            set_seeds(seed, self.env)
+
+    def train(self):
+        raise NotImplementedError
+
+    def save(self):
+        saving_params = self.agent.get_hyperparams()
+        logdir = self.logger.logdir
+        algo = self.agent.__class__.__name__
+        env_name = self.env.envs[0].unwrapped.spec.id
+
+        save_dir = '{}/checkpoints/{}_{}'.format(logdir, algo, env_name)
+        os.makedirs(save_dir, exist_ok=True)
+        torch.save('{}/{}.pt'.format(save_dir, self.ckpt_log_name))
+
+    @property
+    def n_envs(self):
+        return self.env.n_envs
+
+
+class OffPolicyTrainer(Trainer):
+    def __init__(self, agent, env, logger, buffer=None, off_policy=True, save_interval=0,
+                 render=False, max_ep_len=1000, distributed=False, ckpt_log_name='experiment',
+                 steps_per_epoch=4000, epochs=10, device='cpu', log_interval=10, batch_size=50,
+                 warmup_steps=1000, start_update=1000, update_interval=50, seed=0):
+        super(OffPolicyTrainer, self).__init__(agent, env, logger, buffer, 
+                                               off_policy, save_interval, render, max_ep_len,
+                                               distributed, ckpt_log_name,steps_per_epoch, 
+                                               epochs, device, log_interval, batch_size, seed
+                                               )
+        self.warmup_steps = warmup_steps
+        self.update_interval = update_interval
+        self.start_update = start_update
+        
+    def train(self):
+        state, episode_reward, episode_len, episode = self.env.reset(), 0, 0, 0
+        total_steps = self.steps_per_epoch * self.epochs
+
+        if 'noise' in self.agent.__dict__ and self.agent.noise is not None:
+            self.agent.noise.reset()
+
+        for t in range(total_steps):
+
+            if t < self.warmup_steps:
+                action = self.env.action_space.sample()
+            else:
+                action = self.agent.select_action(state)
```

Comment:
We need to be careful here. The argument `deterministic` has to be set according to the algorithm

Reply:
Yeah will change this, I think I removed this when testing SAC.

Reply:
I'll set it up as an arg in the constructor to avoid any assumptions.